nle/dataset/populate_db.py

```diff
@@ -153,38 +199,64 @@ def add_altorg_directory(path, name, filename=db.DB):
 
 
 def add_nledata_directory(path, name, filename=db.DB):
-    with db.db(filename=filename, rw=True) as conn:
+    """This function can be used to add any `nle_data` dataset to a database.
+
+    Full games that are generated by an env such as:
+
+        `gym.make("NetHackChallenge-v0", savedir="", save_ttyrec_every=k)`
+
+    come with the following structure:
+
+    The directory structure should look like:
+
+        nle_data/
+        ├── 20220414-112633_sdfd83ns/
+        │   ├── nle.3968599.0.ttyrec.bz2
+        │   ├── nle.3968599.k.ttyrec.bz2
+        │   └── nle.3968599.xlogfile
+        └── <date>-<time>_<random>/
+            ├── nle.<process-id>.0.ttyrec.bz2
+            ├── nle.<process-id>.k.ttyrec.bz2
+            └── nle.<process-id>.xlogfile
 
+    This algorithm should be deterministic and always return the same dataset
+    from an empty database, regardless of environment.
+    """
+    with db.db(filename=filename, rw=True) as conn:
         logging.info("Adding dataset '%s' ('%s') to '%s' " % (name, path, filename))
+
         root = os.path.abspath(path)
         stime = time.time()
-
         c = conn.cursor()
 
+        # 1. Check if the dataset name exists, and add the root.
         db.create_dataset(name, root, conn=c, commit=False)
 
+        # 2. For each xlogfile, read the games and take only the games that
+        #   correspond to the ttyrecs in the enclosing directory.
         for xlogfile in sorted(glob.iglob(path + "/*/*.xlogfile")):
-
             stem = xlogfile.replace(".xlogfile", ".*.ttyrec.bz2")
+
             resets = set(int(i.split(".")[-3]) for i in glob.iglob(stem))
 
             def filter(gen):
-                # NB: xlogfile may have more rows than in directory
-                #     due to 'save_ttyrec_every' option in env.py
+                # The `xlogfile` may have more rows than files in directory
+                # due to 'save_ttyrec_every' option in env.py, so filter these out.
                 for line_no, line in enumerate(gen):
                     if line_no in resets:
                         yield line
 
+            # 3. Add games to `games` and `datasets` table.
             game_gen = game_data_generator(xlogfile, filter=filter)
             insert_sql = f"""
                 INSERT INTO games
                 VALUES (NULL, {','.join('?' for _ in XLOGFILE_COLUMNS)} )
             """
             c.executemany(insert_sql, game_gen)
-
             gameids = db.get_most_recent_games(c.rowcount, conn=c)
             db.add_games(name, *gameids, conn=conn, commit=False)
 
+            # 4. Add ttyrecs to `ttyrecs` table.
```

Comment:
Does it make sense to allow the "blacklist" ability for these datasets too, even though we don't currently need it (for consistency)?

Reply:
Probably not I think. YAGNI. This is because alt.org is truly a noisy dataset. If there are requests we can add later